Update gcycle to be able to read re-gridded GSI increment land files.

ClaraDraper-NOAA opened the pull request:
## DESCRIPTION OF CHANGES: 

Generalized the reading in of land increments to be: 
1) soil increment OR snow increment file
2) on the GSI output grid to be regridded within global_cycle OR on the native model grid 

Native model grid files may come from JEDI, or be from GSI and been regridded using a separate program. The read routine will detect which.

Also, changed the number of soil levels to be updated from 3 to 2. This will change the test output for all tests adding soil increments.

## TESTS CONDUCTED: 
If there are changes to the build or source code, the tests below must be conducted. Contact a repository manager if you need assistance.

- [x] Compile branch on all Tier 1 machines using Intel (Orion, Jet, Hera, Hercules and WCOSS2). Done using 4d6403f.
- [x] Compile branch on Hera using GNU. Done using 4d6403f. The `global_cycle` code had no warnings.
- [x] Compile branch in 'Debug' mode on WCOSS2. Done on Dogwood using 4d6403f. No warnings.
- [x] Run unit tests locally on any Tier 1 machine. Done on Hera using GNU and 4d6403f. All tests passed.
- [x] Run `global_cycle` consistency tests locally on all Tier 1 machines. Done using 4d6403f. The new test ran to completion with no errors. The other tests all passed as expected.

Describe any additional tests performed.

- Created the Doxygen without warnings using 07ac8bd.

## DEPENDENCIES:
None.

## DOCUMENTATION:
N/A

## ISSUE: 
Fixes #1008

ClaraDraper-NOAA committed:
Updates to generalize inputs land increments to Gaussian grid and fv3 grid.

ClaraDraper-NOAA committed:
Merge branch 'develop' into feature/gcycle_fv3incr

ClaraDraper-NOAA committed:
Udpated namelist variables in gcycle tests.

ClaraDraper-NOAA commented:
@GeorgeGayno-NOAA Submitting this as a draft for now, while I do the tests. I'm not sure which are the unit tests and which are the consistency tests?

I'm currently running rt.sh, with all the tests commented out except global_cycle.
I expect it to be zero-diff without the lsoil_incr change in global_cycle.sh, so am checking that, then will change the lsoil_incr. 
I'll check it compiles with gnu on hera.
What other test should I do? ( note: I don't have access to any machines other than hera and gaeia)

GeorgeGayno-NOAA commented:
> @GeorgeGayno-NOAA Submitting this as a draft for now, while I do the tests. I'm not sure which are the unit tests and which are the consistency tests?
> 
> I'm currently running rt.sh, with all the tests commented out except global_cycle. I expect it to be zero-diff without the lsoil_incr change in global_cycle.sh, so am checking that, then will change the lsoil_incr. I'll check it compiles with gnu on hera. What other test should I do? ( note: I don't have access to any machines other than hera and gaeia)

The regression tests are in `./reg_tests/global_cycle`. To run just the global_cycle test, you can invoke `./driver.hera.sh` from the command line with no arguments. You don't need to bother with the `./rt.sh` script.

The unit tests may be run by some slight modifications to the `./build_all.sh` script:

```
 # The unit test data download is part of the build system. Not all machines can
 # access the EMC ftp site, so turn off the build (-DBUILD_TESTING=OFF) of the units tests accordingly.
 # Those with access to the EMC ftp site are: Orion and Hera.
-CMAKE_FLAGS+=" -DBUILD_TESTING=${BUILD_TESTING:-OFF}"
+CMAKE_FLAGS+=" -DBUILD_TESTING=${BUILD_TESTING:-ON}"
```

You may also need to update `./cmake/mpiexec.hera` for an ACCOUNT you have access to.

ClaraDraper-NOAA committed:
update reg_test namelist to use new variables.

ClaraDraper-NOAA committed:
Added LSOIL_INCR variable.

ClaraDraper-NOAA commented:
All 5 consistency checks passed on hera. 
Compiled with gnu on hera.

ClaraDraper-NOAA commented:
@GeorgeGayno-NOAA How do I confirm that the ctests passed. The first pass through, they didn't and I fixed it.  build_all.sh compiled, but I can't find a report of the test results.

ClaraDraper-NOAA committed:
Update to test for new read_data arguments.

GeorgeGayno-NOAA commented:
> @GeorgeGayno-NOAA How do I confirm that the ctests passed. The first pass through, they didn't and I fixed it. build_all.sh compiled, but I can't find a report of the test results.

When I test your branch on Hera, I see one unit test failing:

```
      Start 26: global_cycle-ftst_read_increments
26/38 Test #26: global_cycle-ftst_read_increments ...........***Failed    3.74 sec
```

The test results will be in ./build/Testing/Temporary/LastTest.log.  The test fails because of a missing file:

```
 FATAL ERROR: OPENING FILE: ./fnbgsi.001: No such file or directory
 STOP.
```

I can take a look.

ClaraDraper-NOAA commented:
@GeorgeGayno-NOAA Is that from standard out? I think I'm not actually running the tests. Let me follow up on that.

GeorgeGayno-NOAA commented:
> @GeorgeGayno-NOAA Is that from standard out? I think I'm not actually running the tests. Let me follow up on that.

That is the output from the `build_all.sh` script after setting `BUILD_TESTING=ON` and uncommenting 'ctest' at the bottom.

ClaraDraper-NOAA committed:
Fixed filename in ctest for soil increment.

ClaraDraper-NOAA commented:
I fixed it. The reg_tests and ctests pass on hera, and gnu compilation works on hera. Converting from a draft. Yuan Xue and Tseganeh Gichamo would be good reviewers.

GeorgeGayno-NOAA reviewed (commented):
Do we need another regression test for a GSI case on the native grid?

yuanxue2870 commented on sorc/global_cycle.fd/cycle.f90:
```diff
@@ -74,10 +65,10 @@
-!!  -DO_SNO_INC_JEDI    Do land increments to snow states on cubed-sphere tiles
-!!                      (Noah land model only).
+!!  -DO_SOI_INC    Do land increments to soil states
+!!  -INTERP_LNDINC Land increment us from GSI and needs to be regridded
```
```suggestion
!!  -INTERP_LNDINC Land increment is from GSI and needs to be regridded
```

ClaraDraper-NOAA replied:
Done

yuanxue2870 commented:
Overall, looks good to me. Thanks!

GeorgeGayno-NOAA commented:
@ClaraDraper-NOAA - make sure you have merged the latest updates from 'develop'.

tsga reviewed (approved):
Looks good to me. 
@ClaraDraper-NOAA I am guessing the "gaussian_to_fv3_interp" function will soon be replaced by the ESMF re-gridding code you just wrote?

GeorgeGayno-NOAA commented:
> Do we need another regression test for a GSI case on the native grid?

To expand on my comment, there are two regression tests for applying land increments: `C192.jedi_lndincsoilnoahmp.sh` and `C192.gsi_lndincsoilnoahmp.sh`. The latter uses GSI data that needs to be re-gridded to the tiles. Should a test using tiled GSI data (no re-gridding necessary) be added? If you have a test case using tiled GSI data, I would be happy to convert it into a test.

GeorgeGayno-NOAA commented:
Is logical `do_lndinc` really needed? Are `do_soi_inc` and `do_sno_inc` sufficient?

ClaraDraper-NOAA committed:
Merge branch 'develop' into feature/gcycle_fv3incr

ClaraDraper-NOAA commented:
> @ClaraDraper-NOAA - make sure you have merged the latest updates from 'develop'.

done.

GeorgeGayno-NOAA commented:
> > > Do we need another regression test for a GSI case on the native grid?
> > 
> > 
> > To expand on my comment, there are two regression tests for applying land increments: `C192.jedi_lndincsoilnoahmp.sh` and `C192.gsi_lndincsoilnoahmp.sh`. The latter uses GSI data that needs to be re-gridded to the tiles. Should a test using tiled GSI data (no re-gridding necessary) be added? If you have a test case using tiled GSI data, I would be happy to convert it into a test.
> 
> The read routine is slightly different for a re-gridded GSI file than for the two test cases that are already there. If you'd like me to prepare a new test, I can.

Yes. Please do.

ClaraDraper-NOAA committed:
Added new test for adding pre-interpolated increments from GSI to soil states.

ClaraDraper-NOAA commented:
> Looks good to me. @ClaraDraper-NOAA I am guessing the "gaussian_to_fv3_interp" function will soon be replaced by the ESMF re-gridding code you just wrote?

This PR effectively adds the option to do either. However, the ESMF regridding is more appropriate  - so perhaps in the future we can just delete the gaussian_to_fv3_interp as you suggest.

ClaraDraper-NOAA committed:
Tidy up comments.

ClaraDraper-NOAA commented:
> > > > Do we need another regression test for a GSI case on the native grid?
> > > 
> > > 
> > > To expand on my comment, there are two regression tests for applying land increments: `C192.jedi_lndincsoilnoahmp.sh` and `C192.gsi_lndincsoilnoahmp.sh`. The latter uses GSI data that needs to be re-gridded to the tiles. Should a test using tiled GSI data (no re-gridding necessary) be added? If you have a test case using tiled GSI data, I would be happy to convert it into a test.
> > 
> > 
> > The read routine is slightly different for a re-gridded GSI file than for the two test cases that are already there. If you'd like me to prepare a new test, I can.
> 
> Yes. Please do.

I just added a new test, and have checked it passes on hera. I added the new test to the drivers for the other machines, except s4 which looks a bit different to the others. 

The additional files that you need to add to the COMIN are: 
/scratch2/BMC/gsienkf/Clara.Draper/gerrit-hera/PRprep/NEWTEST_UFS_UTILS/UFS_UTILS_COMIN/soil_xainc* 

Examples of the baseline files are here: 
/scratch2/BMC/gsienkf/Clara.Draper/gerrit-hera/PRprep/NEWTEST_UFS_UTILS/NEW_BASELINE_FILES/c192.gsitile_lndincsoilnoahmp

ClaraDraper-NOAA commented:
> Is logical `do_lndinc` really needed? Are `do_soi_inc` and `do_sno_inc` sufficient?

@GeorgeGayno-NOAA Yes, do_lndinc is redundant, since it can be turned on by the do_*_inc options. However, my preference would be to keep it (together with the DO_SFCCYCLE and DONST) to make it clear which actions (adding land increments, updating land model parameters, and/or updating NST) cycle if being called for. It's up to you though - if you'd like me to remove it, let me know. We could either completely remove it, or remove it from the namelists, and set it internally.

ClaraDraper-NOAA commented:
@GeorgeGayno-NOAA I think I've addressed everything, let me know if there's anything else I can do.

GeorgeGayno-NOAA commented:
> > Is logical `do_lndinc` really needed? Are `do_soi_inc` and `do_sno_inc` sufficient?
> 
> @GeorgeGayno-NOAA Yes, do_lndinc is redundant, since it can be turned on by the do_*_inc options. However, my preference would be to keep it (together with the DO_SFCCYCLE and DONST) to make it clear which actions (adding land increments, updating land model parameters, and/or updating NST) cycle if being called for. It's up to you though - if you'd like me to remove it, let me know. We could either completely remove it, or remove it from the namelists, and set it internally.

Ok. We can keep it. But I would update the script and code prologs, so the general user knows how to set each variable. For example, I am guessing if 'do_soi_inc' is set to true, then 'do_lndinc' must also be set to true.

GeorgeGayno-NOAA commented on ush/global_cycle.sh:
```diff
@@ -141,9 +141,11 @@
-#     DO_SOI_INC_GSI    Call routine to update soil states with gsi(gaussian) increment files
-#     DO_SNO_INC_JEDI   Call routine to update snow states with jedi increment files
-#     DO_SOI_INC_JEDI   Call routine to update soil states with jedi increment files
+#     DO_SOI_INC    Call routine to update soil states 
```
I would add more detail on how to set each logical.

ClaraDraper-NOAA replied:
I've updated the comments, and added some logic into the global_driver.sh to set DO_LNDINC to true if either of DO_SNO_INC or DO_SOI_INC it true.

GeorgeGayno-NOAA commented:
> > > > > Do we need another regression test for a GSI case on the native grid?
> > > > 
> > > > 
> > > > To expand on my comment, there are two regression tests for applying land increments: `C192.jedi_lndincsoilnoahmp.sh` and `C192.gsi_lndincsoilnoahmp.sh`. The latter uses GSI data that needs to be re-gridded to the tiles. Should a test using tiled GSI data (no re-gridding necessary) be added? If you have a test case using tiled GSI data, I would be happy to convert it into a test.
> > > 
> > > 
> > > The read routine is slightly different for a re-gridded GSI file than for the two test cases that are already there. If you'd like me to prepare a new test, I can.
> > 
> > 
> > Yes. Please do.
> 
> I just added a new test, and have checked it passes on hera. I added the new test to the drivers for the other machines, except s4 which looks a bit different to the others.
> 
> The additional files that you need to add to the COMIN are: /scratch2/BMC/gsienkf/Clara.Draper/gerrit-hera/PRprep/NEWTEST_UFS_UTILS/UFS_UTILS_COMIN/soil_xainc*
> 
> Examples of the baseline files are here: /scratch2/BMC/gsienkf/Clara.Draper/gerrit-hera/PRprep/NEWTEST_UFS_UTILS/NEW_BASELINE_FILES/c192.gsitile_lndincsoilnoahmp

Great. I will host these files in the official location on each machine.

GeorgeGayno-NOAA commented:
I placed the new baseline files on Hera:

```
[role.ufsutils@hfe11 c192.gsitile_lndincsoilnoahmp]$ pwd
/scratch1/NCEPDEV/nems/role.ufsutils/ufs_utils/reg_tests/global_cycle/baseline_data/c192.gsitile_lndincsoilnoahmp
[role.ufsutils@hfe11 c192.gsitile_lndincsoilnoahmp]$ ls -l
total 223104
-r--r--r-- 1 role.ufsutils nems 38073028 Jan 27 16:08 20190730.000000.sfcanl_data.tile1.nc
-r--r--r-- 1 role.ufsutils nems 38073028 Jan 27 16:08 20190730.000000.sfcanl_data.tile2.nc
-r--r--r-- 1 role.ufsutils nems 38073028 Jan 27 16:08 20190730.000000.sfcanl_data.tile3.nc
-r--r--r-- 1 role.ufsutils nems 38073028 Jan 27 16:08 20190730.000000.sfcanl_data.tile4.nc
-r--r--r-- 1 role.ufsutils nems 38073028 Jan 27 16:08 20190730.000000.sfcanl_data.tile5.nc
-r--r--r-- 1 role.ufsutils nems 38073028 Jan 27 16:08 20190730.000000.sfcanl_data.tile6.nc
```

and the new input files here:

```
/scratch1/NCEPDEV/nems/role.ufsutils/ufs_utils/reg_tests/global_cycle/input_data_noahmp
[role.ufsutils@hfe11 input_data_noahmp]$ ls -l soil_xainc.*
-rw-r--r-- 1 role.ufsutils nems 1187094 Jan 27 15:59 soil_xainc.gsi.tile1.nc.pr1010
-rw-r--r-- 1 role.ufsutils nems 1187094 Jan 27 15:59 soil_xainc.gsi.tile2.nc.pr1010
-rw-r--r-- 1 role.ufsutils nems 1187094 Jan 27 15:59 soil_xainc.gsi.tile3.nc.pr1010
-rw-r--r-- 1 role.ufsutils nems 1187094 Jan 27 15:59 soil_xainc.gsi.tile4.nc.pr1010
-rw-r--r-- 1 role.ufsutils nems 1187094 Jan 27 15:59 soil_xainc.gsi.tile5.nc.pr1010
-rw-r--r-- 1 role.ufsutils nems 1187094 Jan 27 15:59 soil_xainc.gsi.tile6.nc.pr1010
lrwxrwxrwx 1 role.ufsutils nems      30 Jan 27 16:03 soil_xainc.tile1.nc -> soil_xainc.gsi.tile1.nc.pr1010
lrwxrwxrwx 1 role.ufsutils nems      30 Jan 27 16:03 soil_xainc.tile2.nc -> soil_xainc.gsi.tile2.nc.pr1010
lrwxrwxrwx 1 role.ufsutils nems      30 Jan 27 16:03 soil_xainc.tile3.nc -> soil_xainc.gsi.tile3.nc.pr1010
lrwxrwxrwx 1 role.ufsutils nems      30 Jan 27 16:04 soil_xainc.tile4.nc -> soil_xainc.gsi.tile4.nc.pr1010
lrwxrwxrwx 1 role.ufsutils nems      30 Jan 27 16:04 soil_xainc.tile5.nc -> soil_xainc.gsi.tile5.nc.pr1010
lrwxrwxrwx 1 role.ufsutils nems      30 Jan 27 16:04 soil_xainc.tile6.nc -> soil_xainc.gsi.tile6.nc.pr1010
```

I will push the data to the other machines after we verify everything is working.

ClaraDraper-NOAA committed:
Clarified do_lndinc, do_soi_inc, and do_sno_inc in comments and script.

ClaraDraper-NOAA committed:
updated INTEPR_LNDINC to GCYCLE_INTERP_LNDINC

ClaraDraper-NOAA commented:
@GeorgeGayno-NOAA Apologies for the poor PR behaviour. I just pushed a minor change to a variable name in the scripts (INTERP_LNDINC to GCYCLE_INTERP_LNDINC), since this makes the function clearer in global_workflow. The tests pass on hera with this change.

GeorgeGayno-NOAA commented on reg_tests/global_cycle/C192.gsitile_lndincsoilnoahmp.sh:
```diff
@@ -0,0 +1,88 @@
+  if [ -f $files ]; then
+    echo CHECK $files
+    $NCCMP -dmfqS $files $HOMEreg/baseline_data/c192.gsi_lndincsoilnoahmp/$files
+    iret=$?
```
The baseline directory should be:
`$HOMEreg/baseline_data/C192.gsitile_lndincsoilnoahmp`

ClaraDraper-NOAA replied:
done

GeorgeGayno-NOAA commented on reg_tests/global_cycle/C192.gsitile_lndincsoilnoahmp.sh:
```diff
@@ -0,0 +1,88 @@
+  echo "**********************************************"
+  if [ "$UPDATE_BASELINE" = "TRUE" ]; then
+    $HOMEgfs/reg_tests/update_baseline.sh $HOMEreg "c192.lndincsoilnoahmp" $commit_num
+  fi
```
The script argument should be `c192.gsitile_lndincsoilnoahmp`

ClaraDraper-NOAA replied:
done

GeorgeGayno-NOAA commented:
I found a bug in `C192.gsi_lndincsoilnoahmp.sh` that is not related to this issue. But I would like to fix it:

```
/scratch1/NCEPDEV/da/George.Gayno/ufs_utils.git/UFS_UTILS.clara/reg_tests/global_cycle $ git diff C192.gsi_lndincsoilnoahmp.sh
--- a/reg_tests/global_cycle/C192.gsi_lndincsoilnoahmp.sh
+++ b/reg_tests/global_cycle/C192.gsi_lndincsoilnoahmp.sh
@@ -88,7 +88,7 @@ if [ $test_failed -ne 0 ]; then
   echo "<<< C192 GSI based LANDINC SOIL-NOAHMP CYCLE TEST FAILED. >>>"
   echo "**********************************************"
   if [ "$UPDATE_BASELINE" = "TRUE" ]; then
-    $HOMEgfs/reg_tests/update_baseline.sh $HOMEreg "c192.lndincsoilnoahmp" $commit_num
+    $HOMEgfs/reg_tests/update_baseline.sh $HOMEreg "c192.gsi_lndincsoilnoahmp" $commit_num
   fi
 else
   echo
```

GeorgeGayno-NOAA commented on ush/global_cycle_driver.sh:
```diff
@@ -95,15 +95,15 @@ for n in $(seq 1 $ntiles); do
   fi
 
-  if [[ "$DO_SOI_INC_GSI" == ".true." ]] ; then
+  if [ "$DO_SOI_INC" == ".true." ] && [ "$GCYCLE_INTERP_LNDINC" == ".true." ] ; then
```
When I run the new regression test, it points to the `soil_sfcincr_jedi` files. There is no logic to pick the tiled gsi data.

ClaraDraper-NOAA replied:
Sorry about that @GeorgeGayno-NOAA. I've fixed it, which meant replacing the new baseline results. I also fixed an issue with the new gsi_tile increments files, and have re-named them slightly.

ClaraDraper-NOAA committed:
addressed George's PR, fixed broken reg-test.

ClaraDraper-NOAA commented:
> I found a bug in `C192.gsi_lndincsoilnoahmp.sh` that is not related to this issue. But I would like to fix it:
> 
> ```
> /scratch1/NCEPDEV/da/George.Gayno/ufs_utils.git/UFS_UTILS.clara/reg_tests/global_cycle $ git diff C192.gsi_lndincsoilnoahmp.sh
> --- a/reg_tests/global_cycle/C192.gsi_lndincsoilnoahmp.sh
> +++ b/reg_tests/global_cycle/C192.gsi_lndincsoilnoahmp.sh
> @@ -88,7 +88,7 @@ if [ $test_failed -ne 0 ]; then
>    echo "<<< C192 GSI based LANDINC SOIL-NOAHMP CYCLE TEST FAILED. >>>"
>    echo "**********************************************"
>    if [ "$UPDATE_BASELINE" = "TRUE" ]; then
> -    $HOMEgfs/reg_tests/update_baseline.sh $HOMEreg "c192.lndincsoilnoahmp" $commit_num
> +    $HOMEgfs/reg_tests/update_baseline.sh $HOMEreg "c192.gsi_lndincsoilnoahmp" $commit_num
>    fi
>  else
>    echo
> ```

I fixed this.

ClaraDraper-NOAA commented:
@GeorgeGayno-NOAA I've addressed everything.

GeorgeGayno-NOAA commented:
The `C192.gsitile_lndincsoilnoahmp.sh` test is failing. It is looking for a layer 2 slc_inc record.

```
readingslc2_inc

 FATAL ERROR: READING slc*_inc ID: NetCDF: Variable not found
```

The input files are here: `/scratch1/NCEPDEV/nems/role.ufsutils/ufs_utils/reg_tests/global_cycle/input_data_noahmp/soil_xainc.???` These files have 3 layers of soilt_inc, but only 1 layer of slc_incr. The test sets `LSOIL_INCR=3` **Did I baseline the wrong input files?**

ClaraDraper-NOAA commented:
> The `C192.gsitile_lndincsoilnoahmp.sh` test is failing. It is looking for a layer 2 slc_inc record.
> 
> ```
> readingslc2_inc
> 
>  FATAL ERROR: READING slc*_inc ID: NetCDF: Variable not found
> ```
> 
> The input files are here: `/scratch1/NCEPDEV/nems/role.ufsutils/ufs_utils/reg_tests/global_cycle/input_data_noahmp/soil_xainc.???` These files have 3 layers of soilt_inc, but only 1 layer of slc_incr. The test sets `LSOIL_INCR=3` **Did I baseline the wrong input files?**

I updated to files to fix this. The new ones are here: 

/scratch2/BMC/gsienkf/Clara.Draper/gerrit-hera/PRprep/NEWTEST_UFS_UTILS/UFS_UTILS_COMIN $ ncdump -c soil_xainc.006
netcdf soil_xainc {
dimensions:
	xaxis_1 = 192 ;
	yaxis_1 = 192 ;
variables:
	double soilt1_inc(yaxis_1, xaxis_1) ;
	double soilt2_inc(yaxis_1, xaxis_1) ;
	double soilt3_inc(yaxis_1, xaxis_1) ;
	double slc1_inc(yaxis_1, xaxis_1) ;
	double slc2_inc(yaxis_1, xaxis_1) ;
	double slc3_inc(yaxis_1, xaxis_1) ;
data:
}

GeorgeGayno-NOAA commented:
@ClaraDraper-NOAA - 'develop' was updated on Friday. Please merge these changes to your branch. They should not conflict with your PR.

ClaraDraper-NOAA committed:
Merge branch 'develop' into feature/gcycle_fv3incr

ClaraDraper-NOAA commented:
> @ClaraDraper-NOAA - 'develop' was updated on Friday. Please merge these changes to your branch. They should not conflict with your PR.

done.
